### PR TITLE
Shorten links

### LINKED ISSUE #472: Shorten links to files in comments (WIP)
Look at these beautiful autolinking:

* https://github.com/sindresorhus/refined-github/commit/784a028af8c04da116e0bd5c0c07569eeef52486  
  it was `https://github.com/sindresorhus/refined-github/commit/784a028af8c04da116e0bd5c0c07569eeef52486`
* https://github.com/rollup/rollup/issues/1435  
  it was `https://github.com/rollup/rollup/issues/1435`

But file links don't get the same treatment:

* https://github.com/sindresorhus/refined-github/blob/master/icon.ai
* https://github.com/sindresorhus/refined-github/blob/784a028af8c04da116e0bd5c0c07569eeef52486/src/content.js
* https://github.com/rollup/rollup/blob/master/.eslintrc
* https://github.com/rollup/rollup/blob/rollup-init/.eslintrc
* https://github.com/sindresorhus/refined-github/blame/patch-diff-shortcuts/.editorconfig
* https://raw.githubusercontent.com/sindresorhus/refined-github/patch-diff-shortcuts/readme.md

They should be shortened to:

* [`master`/icon.ai](https://github.com/sindresorhus/refined-github/blob/master/icon.ai)
* [`784a028`/src/content.js](https://github.com/sindresorhus/refined-github/blob/784a028af8c04da116e0bd5c0c07569eeef52486/src/content.js)
* [rollup/rollup/`master`/.eslintrc](https://github.com/rollup/rollup/blob/master/.eslintrc)
* [rollup/rollup/`rollup-init`/.eslintrc](https://github.com/rollup/rollup/blob/rollup-init/.eslintrc)
* [blame/`patch-diff-shortcuts`/.editorconfig](https://github.com/sindresorhus/refined-github/blame/patch-diff-shortcuts/.editorconfig)
* [`patch-diff-shortcuts`/`readme.md`](https://raw.githubusercontent.com/sindresorhus/refined-github/patch-diff-shortcuts/readme.md) <-- link to `raw`

They might look a bit cryptic at first but perhaps they're better than multi-line full URLs. At the very least we can drop `https` (but not `http`), drop `github.com`, shorten commit hashes, and `highlight` branches and commits

### PR DESCRIPTION
**Pull request by bfred-it**
Closes #472. This post serves as test. Most likes are 404s or rick rolls, don't click.

## Repo root

https://github.com/sindresorhus/refined-github/
https://github.com/sindresorhus/refined-github/tree/v0.12
https://github.com/sindresorhus/refined-github/tree/d71718db6aa4feb8dc10edbad1134472468e971a

https://github.com/nodejs/node/
https://github.com/nodejs/node/tree/v0.12
https://github.com/nodejs/node/tree/d71718db6aa4feb8dc10edbad1134472468e971a

## Directory

https://github.com/sindresorhus/refined-github/tree/master/doc
https://github.com/sindresorhus/refined-github/tree/v0.12/doc
https://github.com/sindresorhus/refined-github/tree/d71718db6aa4feb8dc10edbad1134472468e971a/doc

https://github.com/nodejs/node/tree/master/doc
https://github.com/nodejs/node/tree/v0.12/doc
https://github.com/nodejs/node/tree/d71718db6aa4feb8dc10edbad1134472468e971a/doc

## File

https://github.com/sindresorhus/refined-github/blob/master/.gitignore
https://github.com/sindresorhus/refined-github/blob/v0.12/.gitignore
https://github.com/sindresorhus/refined-github/blob/cc8fc46/.gitignore

https://github.com/nodejs/node/blob/master/.gitignore
https://github.com/nodejs/node/blob/v0.12/.gitignore
https://github.com/nodejs/node/blob/cc8fc46/.gitignore

## File blame

https://github.com/sindresorhus/refined-github/blame/master/.gitignore
https://github.com/sindresorhus/refined-github/blame/v0.12/.gitignore
https://github.com/sindresorhus/refined-github/blame/cc8fc46/.gitignore

https://github.com/nodejs/node/blame/master/.gitignore
https://github.com/nodejs/node/blame/v0.12/.gitignore
https://github.com/nodejs/node/blame/cc8fc46/.gitignore

## File commits/history

https://github.com/sindresorhus/refined-github/commits/master/.gitignore
https://github.com/sindresorhus/refined-github/commits/v0.12/.gitignore
https://github.com/sindresorhus/refined-github/commits/cc8fc46/.gitignore

https://github.com/nodejs/node/commits/master/.gitignore
https://github.com/nodejs/node/commits/v0.12/.gitignore
https://github.com/nodejs/node/commits/cc8fc46/.gitignore

## Commit diff and patch

https://github.com/sindresorhus/refined-github/commit/cc8fc46.diff
https://github.com/sindresorhus/refined-github/commit/cc8fc46.patch

https://github.com/nodejs/node/commit/cc8fc46.diff
https://github.com/nodejs/node/commit/cc8fc46.patch

## Tag/release

https://github.com/sindresorhus/refined-github/releases/tag/v0.12.0

https://github.com/nodejs/node/releases/tag/v0.12.0

## Milestones

https://github.com/sindresorhus/refined-github/milestone/25

https://github.com/nodejs/node/milestone/25

## Labels

https://github.com/sindresorhus/refined-github/labels/npm

https://github.com/nodejs/node/labels/npm

## Downloads

https://github.com/sindresorhus/refined-github/archive/6.4.1.zip
https://github.com/sindresorhus/refined-github/releases/download/6.4.1/now-macos

https://github.com/zeit/now-cli/archive/6.4.1.zip
https://github.com/zeit/now-cli/releases/download/6.4.1/now-macos

## Repo pages

https://github.com/sindresorhus/refined-github/wiki
https://github.com/sindresorhus/refined-github/pulse
https://github.com/sindresorhus/refined-github/labels
https://github.com/sindresorhus/refined-github/network
https://github.com/sindresorhus/refined-github/projects
https://github.com/sindresorhus/refined-github/releases
https://github.com/sindresorhus/refined-github/milestones
https://github.com/sindresorhus/refined-github/contributors

https://github.com/nodejs/node/wiki
https://github.com/nodejs/node/pulse
https://github.com/nodejs/node/labels
https://github.com/nodejs/node/network
https://github.com/nodejs/node/projects
https://github.com/nodejs/node/releases
https://github.com/nodejs/node/milestones
https://github.com/nodejs/node/contributors
https://github.com/nodejs/node/graphs/commit-activity
etc...

## Raw links

https://rawgit.com/sindresorhus/refined-github/master/.gitignore
https://cdn.rawgit.com/sindresorhus/refined-github/v0.12/.gitignore
https://cdn.rawgit.com/sindresorhus/refined-github/d71718db/.gitignore
https://raw.githubusercontent.com/sindresorhus/refined-github/master/.gitignore
https://raw.githubusercontent.com/sindresorhus/refined-github/v0.12/.gitignore
https://raw.githubusercontent.com/sindresorhus/refined-github/d71718db/.gitignore

https://rawgit.com/nodejs/node/master/.gitignore
https://cdn.rawgit.com/nodejs/node/v0.12/.gitignore
https://cdn.rawgit.com/nodejs/node/d71718db/.gitignore
https://raw.githubusercontent.com/nodejs/node/master/.gitignore
https://raw.githubusercontent.com/nodejs/node/v0.12/.gitignore
https://raw.githubusercontent.com/nodejs/node/d71718db/.gitignore

## Users and orgs

https://github.com/sindresorhus
https://github.com/nodejs

## Reserved URLs

https://github.com/pulls
https://github.com/issues
https://github.com/trending
https://github.com/features
https://github.com/marketplace
https://github.com/trending/developers
https://github.com/settings/profile
etc...

## Non-local URLs

https://www.npmjs.com
https://www.npmjs.com/package/node
https://example.com/nodejs/node/blob/cc8fc46/.gitignore looks like file, but isn't safe to shorten

## Natively shortened

https://github.com/sindresorhus/refined-github/pull/1 PR
https://github.com/sindresorhus/refined-github/issues/3 issue
https://github.com/sindresorhus/refined-github/commit/9d7895c567e8f38abfff35da1b6d6d6a0a06f9aa commit
https://github.com/sindresorhus/refined-github/commit/9d7895c567e8f38abfff35da1b6d6d6a0a06f9aa?w=1 commit with query string
https://github.com/sindresorhus/refined-github/compare/v0.12.0...v0.12.1 compare across branches
https://github.com/sindresorhus/refined-github/compare/v7.x...v8:master compare across forks

https://github.com/nodejs/node/pull/1 PR
https://github.com/nodejs/node/issues/3 issue
https://github.com/nodejs/node/commit/9d7895c567e8f38abfff35da1b6d6d6a0a06f9aa commit
https://github.com/nodejs/node/commit/9d7895c567e8f38abfff35da1b6d6d6a0a06f9aa?w=1 commit with query string
https://github.com/nodejs/node/compare/v0.12.0...v0.12.1 compare across branches
https://github.com/nodejs/node/compare/v7.x...v8:master compare across forks

### CODE REVIEW

**Comment by hkdobrev** [comment]
I'd prefer to use colon as a separator between branch and file path as Git and scp do.

**Comment by bfred-it** [comment]
@hkdobrev I'm not familiar, can you show me?

File: src/content.js [hunk 1]
```diff
@@ -471,6 +471,24 @@ function addTitleToEmojis() {
 	}
 }
 
+function shortenVisibleUrls() {
+	$('a[href]').get().forEach(a => {
```
Comment (sindresorhus) [inline src/content.js]: Prefer a `for-of` loop whenever possible as it's faster and more readable. Just use `continue` instead of `return`.

**Comment by hkdobrev** [comment]
@bfred-it On the final visualisation of the links I recommend the following:

- `https://github.com/vendor/repo/blob/branch/path/to/file.js` -> [`master:path/to/file.js`](https://example.com)
- `https://github.com/vendor/repo/tree/branch/path/to/directory` -> [`master:path/to/directory`](https://example.com)

Example of Git using such notation: https://stackoverflow.com/a/338470/679227
(will link to official manual when it stops giving 500 error pages)

I think this is good because Git is using `:` to separate refs and path specs in various cases, but other software like `scp` is using it as well. Also, the forward slash character `/` is used for both directory separator and often used in the name of branches in different Git workflows.

On the regex matching:

You're trying to do positive lookbehind which could be achieved with a regex like the following:

```
/(?<=2)red/
```

However, positive lookbehind is not supported in the JavaScript regex engine. That's why you'd need to use match groups to extract the text you need. Relevant SO answer: https://stackoverflow.com/a/3569116/679227

E.g. you could use `/2([a-z_-\/\+)/` and then use the first matched group.

**Comment by bfred-it** [comment]
@hkdobrev how about external repos? Is this ok? [user/not-this-repo/`master`:index.js](#no-link)

**Comment by bfred-it** [comment]
I worked around the regex complexity by using URL parsing and explicit variables. This should be safer and it's also more mindful of external URLs.

Also added `raw` handling and switched to `:` after the branch.

but I'm not a fan of how `:` looks, I prefer the first one:

1. [user/repo/`branch`/.gitignore](#the-good) 
2. [user/repo/`branch`:.gitignore](#the-bad-and-the-ugly)

**Comment by hkdobrev** [comment]
If you use the forward slash it's not so easy to determine where the branch name end and where the filename starts. Here, you have used inline code to distinguish it, but on the command line this is not possible and that's why colon is used. I'd prefer to wrap both the branch and the path to a single inline code element.

**Comment by bfred-it** [comment]
The command line does not highlight the branch name though; a monospaced filename can get long too. I don't think it gets lost if only the branch is highlighted.

1.  [user/repo/`branch`/.gitignore](#the-good) 
2.  [user/repo/`branch:.gitignore`](#the-good) <- still not a fan of `:.`

**Comment by jgierer12** [comment]
I'm not a fan of the slash either. What about using `@` similar to npm:
* [icon.ai@`master`](https://github.com/sindresorhus/refined-github/blob/master/icon.ai)
* [src/content.js@`784a028`](https://github.com/sindresorhus/refined-github/blob/784a028af8c04da116e0bd5c0c07569eeef52486/src/content.js)
* [.eslintrc@rollup/rollup/`master`](https://github.com/rollup/rollup/blob/master/.eslintrc)
* [.eslintrc@rollup/rollup/`rollup-init`](https://github.com/rollup/rollup/blob/rollup-init/.eslintrc)
* [*`blame:`* .editorconfig@`patch-diff-shortcuts`](https://github.com/sindresorhus/refined-github/blame/patch-diff-shortcuts/.editorconfig)
* [`readme.md`@`patch-diff-shortcuts`](https://raw.githubusercontent.com/sindresorhus/refined-github/patch-diff-shortcuts/readme.md) <-- link to `raw`

**Comment by bfred-it** [comment]
That's not bad. Perhaps the blame link can be formatted like GitHub's comment links, live example:

https://github.com/sindresorhus/refined-github/pull/473#issue-233840042

Which for blame would be [.editorconfig@`patch-diff-shortcuts` (blame)](https://github.com/sindresorhus/refined-github/blame/patch-diff-shortcuts/.editorconfig)

**Comment by jgierer12** [comment]
Some thoughts:

---

Putting `blame` in brackets definitely seems nicer. I also noticed that GitHub puts the repo before the @ as seen in this comment: https://github.com/sindresorhus/refined-github/pull/467#issuecomment-307973734
So maybe we should do that too:
* [rollup/rollup/.eslintrc@`master`](https://github.com/rollup/rollup/blob/master/.eslintrc)
* [rollup/rollup/.eslintrc@`rollup-init`](https://github.com/rollup/rollup/blob/rollup-init/.eslintrc)
* [rollup/rollup/`.eslintrc`@`rollup-init`](https://raw.githubusercontent.com/rollup/rollup/rollup-init/.eslintrc) <-- link to `raw`

To differentiate between directories and repos, we could always add a leading slash before the path:
* [/icon.ai@`master`](https://github.com/sindresorhus/refined-github/blob/master/icon.ai)
* [/src/content.js@`784a028`](https://github.com/sindresorhus/refined-github/blob/784a028af8c04da116e0bd5c0c07569eeef52486/src/content.js)
* [/.editorconfig@`patch-diff-shortcuts` (blame)](https://github.com/sindresorhus/refined-github/blame/patch-diff-shortcuts/.editorconfig)
* [/`readme.md`@`patch-diff-shortcuts`](https://raw.githubusercontent.com/sindresorhus/refined-github/patch-diff-shortcuts/readme.md) <-- link to `raw`
* [/src/`content.js`@`master`](https://raw.githubusercontent.com/sindresorhus/refined-github/master/src/content.js) <-- link to `raw`

---

Don't forget `tree` links:
* [/src@`master`](https://github.com/sindresorhus/refined-github/tree/master/src)
* [/@`diff-whitespace-bug`](https://github.com/sindresorhus/refined-github/tree/diff-whitespace-bug)
* [sindresorhus/xo/lib@`9fc3b06`](https://github.com/sindresorhus/xo/tree/9fc3b0677eac477c94af67bc62257bb3d2273877/lib)
* [sindresorhus/xo/@`9fc3b06`](https://github.com/sindresorhus/xo/tree/9fc3b0677eac477c94af67bc62257bb3d2273877) (not to be confused with `https://github.com/sindresorhus/xo/commit/9fc3b0677eac477c94af67bc62257bb3d2273877` which GitHub transforms to https://github.com/sindresorhus/xo/commit/9fc3b0677eac477c94af67bc62257bb3d2273877)

---

GitHub always displays branch names with a blue background, via the `branch-name` class. Maybe you could do that too?

---

What about dropping the branch name if it's `master`?
* [/icon.ai](https://github.com/sindresorhus/refined-github/blob/master/icon.ai)
* [rollup/rollup/.eslintrc](https://github.com/rollup/rollup/blob/master/.eslintrc)
* [/src/`content.js`](https://raw.githubusercontent.com/sindresorhus/refined-github/master/src/content.js) <-- link to `raw`
* [/.editorconfig (blame)](https://github.com/sindresorhus/refined-github/blame/master/.editorconfig)

**Comment by bfred-it** [comment]
Just now I'm doing a roundup of any links I find (top comment) so that they can be easily discussed and tested.

> GitHub always displays branch names with a blue background, via the branch-name class. Maybe you could do that too?

That'd be nice, but there's no way to tell a branch apart from a tagname or commit:

`https://github.com/nodejs/node/blob/master/.gitignore`
`https://github.com/nodejs/node/blob/v0.12/.gitignore`
`https://github.com/nodejs/node/blob/cc8fc46/.gitignore`

Should they all (master, v0.12, cc8fc46) be blue?

> What about dropping the branch name if it's `master`?

That was the initial idea, but if felt like additional context and there's always [who uses branches other than `master` as the base branch.](https://github.com/sindresorhus/refined-github/pull/455#issuecomment-304892622)

The links to `raw` could use the regular parens syntax: [rollup/rollup/.eslintrc (raw)](#nope)

**Comment by bfred-it** [comment]
@sindresorhus @hkdobrev @jgierer12 please review the top post and let me know if any URLs should be added.

**Comment by jgierer12** [comment]
Missing repo pages: issues, pulls, milestones, labels, pulse, graphs

**Comment by jgierer12** [comment]
> Should they all (master, v0.12, cc8fc46) be blue?

No, I'd prefer the normal, gray tag. I don't really mind if branches aren't blue, it was just something I noticed.

> That was the initial idea, but if felt like additional context and there's always who uses branches other than master as the base branch.

Right, it will just get confusing if the default branch isn't `master`

> The links to raw could use the regular parens syntax as well: rollup/rollup/.eslintrc (raw)

👍

**Comment by bfred-it** [comment]
> pulse, graphs

There are actually a lot more of these but they're all the same, as far as we're concerned, no changes will likely be made other than to `user/repo/pulse`

**Comment by bfred-it** [comment]
How does this look? Try it on the first post in this PR.

Notes:

- All files are marked by `@reference`, this removes ambiguity between
  - `user/repo/file`
  - `user/repo/a-page-like-issues`
- I kept the repo on the left side of `@` to be consistent across all links:
  - `user/repo@branch`
  - `user/repo/file@branch`

Next:

- (How) do we shorten milestones further? Currently `nodejs/node/milestone/25`
- (How) do we shorten releases further? Currently: `nodejs/node/releases/tag/v0.12.0`
- (How) do we shorten labels further? Currently: `nodejs/node/labels/npm`

**Comment by jgierer12** [comment]
Looking great so far!

Like I said, I would add a leading slash to directories/files to distinguish them from repos. e.g.
[/some/dir@`master`]() vs [some/repo@`master`]()

I would also print all of the repo pages with the parens syntax, e.g. [user/repo (wiki)](), [user/repo (projects)]()

For `commits` links, I think it would be better to use the singular (commit): [nodejs/node/.gitignore@`master` (commit)]()

How about 
* [nodejs/node/25 (milestone)]()
* [nodejs/node/`v0.12.0` (release)]()
* [nodejs/node/npm (label)]()

**Comment by bfred-it** [comment]
> For `commits` links, I think it would be better to use the singular (commit): [nodejs/node/.gitignore@`master` (commit)]()

That would make it look like it's a link to a single commit, rather than the list of commits coming up to it. _Perhaps_ `(history)` would be an alternative term.

For everything else: excellent, will do!

**Comment by jgierer12** [comment]
> That would make it look like it's a link to a single commit, rather than the list of commits coming up to it

My bad, I actually thought they were single commits. In that case, `commits` sounds fine

File: src/content.js [hunk 2]
```diff
@@ -465,6 +466,26 @@ function addTitleToEmojis() {
...
+		// .href automatically adds a / to naked origins
+		// so that needs to be tested too
+		if (a.href !== a.textContent && a.href !== `${a.textContent}/`) {
+			continue;
+		}
+
+		const shortened = shortenUrl(a.href);
+
+		// Don't touch the dom if there's nothing to change
+		if (shortened === a.href) {
```
Comment (jgierer12) [inline src/content.js]: Shouldn't this be `a.innerHtml`?

**Comment by sindresorhus** [comment]
This looks amazing 🙌

---

One suggestion:

[sindresorhus/refined-github/commit/cc8fc46.diff](https://github.com/sindresorhus/refined-github/commit/cc8fc46.diff)

Becomes:

[sindresorhus/refined-github@`cc8fc46`.diff](https://github.com/sindresorhus/refined-github/commit/cc8fc46.diff)

Same with `.patch`.

**Comment by bfred-it** [comment]
Yep I was working on that, but the idea was to use 
[sindresorhus/refined-github@`cc8fc46` (diff)](https://github.com/sindresorhus/refined-github/commit/cc8fc46.diff)

File: src/content.js (hunk 2 above)
Comment (bfred-it) [inline src/content.js]: They're the same with the exception that `.href` can contain an extra blackslash. I'll change it to `.textContent`

**Comment by bfred-it** [comment]
Oooooook I think we're there.

Code-wise not 100% happy but I'm already at my third refactoring at this point. 😅

@jgierer12 I didn't shorten the milestone because the `25 (milestone)` wasn't really shorter than `milestone/25`; 25 isn't exactly indicative of its content; 25 looks too much like an issue number.

**Comment by bfred-it** [comment]
@sindresorhus I ended up going for `@cc8fc46.diff` because my suggested `@cc8fc46 (diff)` may look like a link to the diff view rather than a diff file.

File: test/fixtures/window.js [hunk 3]
```diff
@@ -1,20 +1,7 @@
...
-		this.hostname = uriParts.hostname;
-		this.pathname = uriParts.pathname;
-		this._currentURI = uri;
-	},
-
-	get href() {
-		return this._currentURI;
-	}
-};
-
```
Comment (bfred-it) [inline test/fixtures/window.js]: @sindresorhus please verify. This is Node 7+

https://nodejs.org/api/url.html#url_the_whatwg_url_api

Comment (sindresorhus) [inline test/fixtures/window.js]: 👍

**Comment by sindresorhus** [comment]
Can you add it to the feature highlights in the readme with a screenshot?

**Comment by bfred-it** [comment]
screenshot-shortened-urls

**Comment by sindresorhus** [comment]
Perfect

**Comment by bfred-it** [comment]
🎉

**Comment by sindresorhus** [comment]
🍾✨🎉

**Comment by hkdobrev** [comment]
🙌 Great work @bfred-it!

**Comment by bfred-it** [comment]
I'm loving this. Perhaps we should follow @hkdobrev's advice and use : between the repo and filename:

[user/repo:filename@`master`](#nope)
[/filename@`master`](#nope)

**Comment by sindresorhus** [comment]
@bfred-it 👍

**Comment by FernandoMiguel** [comment]
is there a way we can avoid some of these short bits on readme.md?
it's breaking my documentation commands, preventing easy copy paste

for instance:
`git submodule add https://github.com/xxx/Docker.git Dockerfile`
loses the URL so you can execute that

**Comment by bfred-it** [comment]
That should not happen (#789). URL shortening is supposed to only happen to real links. If it's a regression, open a new issue.

**Pull request merged**